setup/helpers/abs_helper.rb

```diff
@@ -0,0 +1,657 @@
+                                     "X-Auth-Token" => abs_token)
+    else
+      puts "Unable to prepare a valid ABS request without a valid token"
+    end
+    req
+  end
+
+  # Prepares and submits the awsdirect request with the specified URI and body
+  #
+  # @author Bill Claytor
+  #
+  # @param [URI] uri The URI where the request should be sent
+  # @param [JSON] body The request body
+  #
+  # @return [Net::HTTPResponse, nil] The HTTP response or nil if unsuccessful
+  #
+  # @example
+  #   response = abs_request_awsdirect(uri, body)
+  #
+  def perform_awsdirect_request(uri, body)
+    res = nil
+    req = get_abs_post_request(uri)
+
+    if req
+      req.body = body
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      http.read_timeout = 120
+
+      puts "sending request body:"
+      puts body
+      puts "to uri: #{uri}"
+      puts
+
+      res = http.request(req)
+      puts "response code: #{res.code}" unless res.nil?
+      puts "response body: #{res.body}" unless res.nil?
+      puts
+    else
+      puts "Unable to complete the specified ABS request"
+    end
+
+    res
+  end
+
+  # Determines whether the ABS response is valid
+  #
+  # @author Bill Claytor
+  #
+  # @param [Net::HTTPResponse] res The HTTP response to evaluate
+  # @param [Array] valid_response_codes The list of valid response codes
+  # @param [Array] invalid_response_bodies The list of invalid response bodies
+  #
+  # @return [true,false] Based on whether the response is valid
+  #
+  # @example
+  #   valid = abs_valid_response?(res, ["200"], ["", nil])
+  #
+  def valid_abs_response?(res, valid_response_codes = ["200"],
+                          invalid_response_bodies = ["", nil])
+    # TODO: other criteria?
+    is_valid_response = false
+
+    if res.nil? || !valid_response_codes.include?(res.code) ||
+       invalid_response_bodies.include?(res.body)
+      puts "Invalid ABS response: "
+      puts "code: #{res.code}" unless res.nil?
+      puts "body: #{res.body}" unless res.nil?
+      puts
+    else
+      is_valid_response = true
+    end
+    is_valid_response
+  end
+
+  # Creates a Beaker-formatted version of the response body
+  # for the provisioned host
+  #
+  # Beaker requires hosts to be specified in this format
+  #
+  # @author Bill Claytor
+  #
+  # @param [Hash] response_body The response body for the provisioned host
+  #
+  # @return [Hash] The updated host
+  #
+  # @example
+  #   host = abs_reformat_resource_host(response_body)
+  #
+  def reformat_abs_resource_host(response_body)
+    reformatted_json = nil
+
+    begin
+      host = JSON.parse(response_body)
+      hostname = host["hostname"]
+
+      new_response_body = {
+        'hostname': hostname,
+        'type':     ABS_BEAKER_TYPE,
+        'engine':   ABS_BEAKER_ENGINE
+      }
+
+      reformatted_json = new_response_body.to_json
+    rescue
+      # TODO: raise?
```

Comment:
We should rescue JSON::ParseError only if that's what we're looking for here, and I think we should re-raise it. If we can't parse the response body then something else downstream will blow up and we won't have this context.

Reply:
Removed and updated get_abs_resource_hosts to rescue and re-raise after returning any provisioned hosts.